### DISCUSSION TITLE
How to create macvlan network in rootless mode?

### DISCUSSION THREAD

**Discussion by ankorth** [moby/moby#44538] [Q&A]
Hello Community,

if you feel TL;DR, go to section "The issue".

this question relates to an attempt to create a macvlan network in rootless mode on a Raspberry OS host with pre-created vlan devices on the host. I believe there is a high "risk" that the answer is "there is just no way to do it", but I'm going to make the question big anyway, because I hope can point me to alternative solutions in such a case. 

Below you will find the following information.

- Environment - output of `docker info` and `docker version`
- Host modifications - config of eth0, eth0.10, and eth0.20 on the host
- Use case - it won't hurt if you know my aim, because there may be a better way
- The issue 

### Output of `docker version`
~~~text
Client: Docker Engine - Community
 Version:           20.10.21
 API version:       1.41
 Go version:        go1.18.7
 Git commit:        baeda1f
 Built:             Tue Oct 25 17:59:27 2022
 OS/Arch:           linux/arm
 Context:           rootless
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.21
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.7
  Git commit:       3056208
  Built:            Tue Oct 25 17:57:19 2022
  OS/Arch:          linux/arm
  Experimental:     false
 containerd:
  Version:          1.6.10
  GitCommit:        770bd0108c32f3fb5c73ae1264f7e503fe7b2661
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
 ~~~

### Output of `docker info`

~~~text
Client:
 Context:    rootless
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.9.1-docker)
  compose: Docker Compose (Docker Inc., v2.12.2)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 2
 Server Version: 20.10.21
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: false
  userxattr: true
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 770bd0108c32f3fb5c73ae1264f7e503fe7b2661
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  rootless
  cgroupns
 Kernel Version: 5.15.30-v7l+
 Operating System: Raspbian GNU/Linux 11 (bullseye)
 OSType: linux
 Architecture: armv7l
 CPUs: 4
 Total Memory: 3.797GiB
 Name: pi4-srv
 ID: 7HDD:JSNO:6ZV2:HXZ4:44GP:SNUZ:2446:G2NO:JF4K:WY2K:ISLI:73GM
 Docker Root Dir: /home/dockuser/.local/share/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
WARNING: No cpu shares support
WARNING: No cpuset support
WARNING: No io.weight support
WARNING: No io.weight (per device) support
WARNING: No io.max (rbps) support
WARNING: No io.max (wbps) support
WARNING: No io.max (riops) support
WARNING: No io.max (wiops) support
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
~~~

### User (rootless)

I created a user `dockuser` which is intended to just run dockerd in rootless mode as a systemd service, and the client. The set up of rootless has been done according to the [rootless guide at docker docs](https://docs.docker.com/engine/security/rootless/) and the advices given in the output of `dockerd-rootless-setuptool.sh`.

### Network
This section may not be so relevant but who knows? Because I didn't know about macvlan driver when I set up the Host system, I created two macvlan interfaces on the host using systemd-networkd, motivated by the use case (see section below). This is done by placing the following files in `/etc/systemd/network/`.

`10-eth0.10.netdev`:

~~~ini
[NetDev]
Name=eth0.10
Kind=macvlan

[MACVLAN]
Mode=bridge
~~~

`11-eth0.20.netdev`:

~~~ini
[NetDev]
Name=eth0.20
Kind=macvlan

[MACVLAN]
Mode=bridge
~~~

`15-eth0.10.network`:

~~~ini
[Match]
Name=eth0.10

[Network]
DHCP=yes

[DHCPv4]
Hostname=openhab
~~~

`16-eth0.20.network`:

~~~ini
[Match]
Name=eth0.20

[Network]
DHCP=yes

[DHCPv4]
Hostname=nextcloud
~~~

`20-eth0.network`:

~~~ini
[Match]
Name=eth0

[Network]
MACVLAN=eth0.10
MACVLAN=eth0.20
~~~

With these files, the output of `ifconfig` shows `eth0`, `eth0.10`, `eth0.20`, and `lo` where each eth gets an individual ip address from my router.

## Use case

The Raspberry shall provide some applications for personal and family use. Some apps shall be available from the web by one sub-directory per app. e.g. `https://rasppi.foobar.xyz/nextcloud`, which would be routed to `https://<ip of eth0>/nextcloud`. Traefik is used route that to the desired docker containers. Some other apps shall only be available from the LAN, e.g. smart home stuff. This is the first reason why I created the secondary interfaces. The second reason is that some containers really don't like if they are accessible from a sub directory. E.G. OpenHab likes to be at `http://openhab` or `http://<ip of eth0.10>` and doesn't like `http://something/openhab`.

I think the macvlan network driver is just intended to be used for that, so I can create one docker network for each individual interface on the host. Furthermore I think that it would unlock the possibility to work with several different Traefik entry points, which would make it possible to distinguish between LAN and Web applications. 

## The issue

If I try to create a macvlan network, for example with 

~~~bash
docker network create -d macvlan --subnet=192.168.1.0/24 --gateway=192.168.1.1 -o parent=eth0.10 macvlan1
~~~ 

This throws the error:

> Error response from daemon: -o parent interface does was not found on the host: eth0

To make sure that I didn't understand something wrong in the docs, I tried the same with a rootful dockerd and that seems to work (I mean, it takes the command and creates the docker network).

What I tried to do so far:

- On the host, I added `dockuser` to the `netdev` group, because I thought that dockusers' dockerd just has no permissions to do anything with the interfaces. That didn't change something.
- I tried to start a rootful dockerd instance and added some command line arguments to point to the runtime directory of the rootless dockerd, but that made the rootless instance inoperable (it damaged something in the config - no surprise).
- I tried to dive into slirp4netns and the namespace stuff related to rootless, because I think it may have something to do with that. But I gave up rather fast because it crosses a border line of efforts I'm able to handle for this project.

**Comment by ankorth** [comment]
The first fact is, that `docker network create -d macvlan […]` works as expected. It is important to realize, that in rootless mode, `dockerd` runs in an isolated namespace. This can be entered with the following command, if one logs in as the user running `dockerd`:

~~~sh
nsenter -U --preserve-credentials -n -t $(pgrep dockerd)
~~~

The output if `ip a` may be something like that:

~~~text
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: tap0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 65520 qdisc pfifo_fast state UP group default qlen 1000
    link/ether e2:47:b7:2a:9e:9e brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.100/24 scope global tap0
       valid_lft forever preferred_lft forever
    inet6 fe80::e047:b7ff:fe2a:9e9e/64 scope link 
       valid_lft forever preferred_lft forever
3: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default 
    link/ether 02:42:6d:ee:0d:a0 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
~~~

`tap0` is the interface provided by [slirp4netns](https://github.com/rootless-containers/slirp4netns). `docker0` is the interface which belongs to the default bridge network.

The docker CLI commands are executed in context of the namespace. Because of that, the `parent` option of `docker network create` (see post above) would have to be filled with an interface which exists in the list, e.g. tap0. But that does not make sense, at least without additional configuration.

Well, no reason to give up. I hope it's possible to meaningfully connect the individual network stacks with each other in a way that a running container is reachable from the LAN with the IP attached to eth0.10.

~~~text
┌──────────────────┐  ┌──────────────┐  ┌───────────────┐
│                  │  │              │  │               │
│ Host network     │  │  Network in  │  │  running      │
│                  │  │  dockerd's   │  │  container    │
│ ┌────┐ ┌───────┐ │  │  namespace   │  │               │
│ │eth0│ │eth0.10│ │  │              │  │               │
│ └────┘ └───────┘ │  │              │  │               │
│             ▲    │  │              │  │               │
│             │    │  │     ┌───┐    │  │               │
│             │    │  │     │ ? │◄───┼──┼─►             │
│     ┌───┐   │    │  │     └───┘    │  │               │
│     │ ? │◄──┘    │  │       ▲      │  │               │
│     └───┘        │  │       │      │  │               │
│       ▲          │  │       │      │  │               │
│       │          │  │       ▼      │  │               │
│       ▼          │  │  ┌─────┐     │  │               │
│    ┌────┐        │  │  │veth-│     │  │               │
│    │veth│◄───────┼──┼─►│peer │     │  │               │
│    └────┘        │  │  └─────┘     │  │               │
│                  │  │              │  │               │
└──────────────────┘  └──────────────┘  └───────────────┘
~~~

Where veth and veth-peer are virtual ethernet devices created in the root namespace.

**Reply by AkihiroSuda** [reply]
> Where veth and veth-peer are virtual ethernet devices created in the root namespace.

`lxc-user-nic` might be of your interest
https://github.com/rootless-containers/rootlesskit/blob/master/docs/network.md#--netlxc-user-nic-experimental

Can be specified here:
https://github.com/moby/moby/blob/66ca24d3a277ba63ce2f51329cd903e0648288fa/contrib/dockerd-rootless.sh#L12

**Comment by redlinejoes** [comment]
@ankorth Were you able to figure this out? I'm trying to solve this same problem. I would like to collaborate on this. I have others who are helping too.

**Reply by ankorth** [reply]
Hi Joseph,

No.

So far I came to the conclusion, that there is not a user friendly way, to implement is. Afterfter spending some further couple of hours of investigation I decided to work on other stuff.

I still believe it's posible to get it working with manually creating interfaces in the running container, the namespace (of rootless docker daemon), and on the host plus NAT rules on host and namespace. But I'm not sure whether something is possible, which restores itself during a restart.

Because of that I'm also analysing whether another way is also worth a try: using a Traefik Container to route Web services and LAN services accordingly.

**Comment by LeLights** [comment]
I am using a VLAN with ID 4, accessed via a tagged interface enp2s0 (enp2s0.4).
I want to expose my rootless Docker container to this VLAN.
The IP I am trying to assign to the container is 172.25.4.192, and it should be able to communicate on the VLAN IP range 172.25.4.0/24.

Create VLAN interface on the host if it does not exist already:

```
(root) ip link add link enp2s0 enp2s0.4 type vlan id 4
(root) ip addr add 172.25.4.192/24 dev enp2s0.4
(root) ip link set enp2s0.4 up
```

Find the pid of the running docker container:

```
(root) machinectl shell rootless_user@
(rootless_user): docker inspect -f '{{.State.Pid}}' container_name
```

Create veth pair

```
(root) ip link add veth-container type veth peer name veth-host
(root) ip link set veth-container netns container_pid
```

Configure veth inside the container namespace:

```
(root) machinectl shell rootless_user@
(rootless_user) nsenter -U --preserve-credentials -n -t container_pid
(container_namespace) ip addr add 172.25.4.192/24 dev veth-container
(container_namespace) ip link set veth-container up
(container_namespace) ip route replace default via 172.25.4.1 dev veth-container
```
Replacing the default gateway is only necessary if responses cannot be handled by the normal default gateway. 172.25.2.19 — for example, if it pings 172.25.4.192 and 172.25.2.19 cannot be reached via the normal default gateway for a response.

Configure veth and bridge on the host:

```
(root) ip addr add 172.25.4.192/24 dev veth-host
(root) ip link set veth-host up
(root) ip link add name br-vlan type bridge
(root) ip link set br-vlan up
(root) ip link set enp2s0.4 master br-vlan
(root) ip link set veth-host master br-vlan
```

Currently, this setup is discarded on reboots, but I think it could be made permanent. (WIP)
Security implications should be considered when using this setup, because Docker’s network restrictions are bypassed for incoming and outgoing traffic, and the VLAN is now used as the default gateway. Hope this helps.
